src/org/omegat/core/machinetranslators/MTHUBTranslate.java

```diff
@@ -0,0 +1,481 @@
+        if (prev != null) {
+            return prev;
+        }
+
+        if (getAvailableLanguageCodes() == null) {
+            return OStrings.getString("MT_ENGINE_MTHUB_INVALID_KEY");
+        }
+
+        String encoding = StandardCharsets.UTF_8.name();
+        try {
+            URL url = new URL(GT_URL);
+            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+
+            conn.setDoOutput(true);
+            conn.setRequestProperty("Accept-Charset", encoding);
+            conn.setRequestProperty("Content-Type",
+                                        "application/json;charset=" + encoding);
+
+            try (OutputStream output = conn.getOutputStream()) {
+
+                List<String> segments = Collections.singletonList(text.substring(0,
+                                 Math.min(text.length(), LIMIT_CHARACTER - 1)));
+
+                MTHUBTranslateRequest mttr = new MTHUBTranslateRequest(
+                        mthubKey, normaliseCode(sLang), normaliseCode(tLang),
+                        segments);
+
+                String json = OM.writeValueAsString(mttr);
+                output.write(json.getBytes(encoding));
+
+                int statusCode = conn.getResponseCode();
+
+                try (InputStream in = (statusCode >= 200 && statusCode < 400)
+                        ? conn.getInputStream() : conn.getErrorStream();) {
+                    MTHUBTranslateResponse response = OM.readValue(
+                            IOUtils.toString(in, StandardCharsets.UTF_8),
+                            MTHUBTranslateResponse.class);
+                    String translation = "";
+                    if (response.isSuccess()) {
+                        translation = response.getData().
+                                      get(0).getSegments().get(0).get(0).
+                                        getTranslation();
+                        putToCache(sLang, tLang, text, translation);
+                    } else {
+                        switch (response.getError().get(0).code) {
+                            case 1: return OStrings.getString("MTHUB_ERROR",
+                                    response.getError().get(0).code,
+                                    OStrings.getString("MT_ENGINE_MTHUB_INVALID_KEY"));
+                            case 3: return OStrings.getString("MTHUB_ERROR",
+                                    response.getError().get(0).code,
+                                    OStrings.getString("MT_ENGINE_MTHUB_INVALID_LANG_CODE"));
+                            default: return OStrings.getString(
+                                        "MTHUB_ERROR", "400",
+                                        response.getError().get(0).getMessage());
+                        }
+                    }
+                    return translation;
+                } catch (Exception e) {
+                    return OStrings.getString("MTHUB_ERROR", "400", e.getMessage());
```

Comment:
These cases where you're passing a string error code like `"400"` and `"404"` seem fishy. How do you know that's the code? If that's not really the code, maybe replace it with `"?"` or something?

Reply:
the code is not known, so I delete it and show only the error message it returns.
OStrings.getString("MTHUB_ERROR", e.getMessage());

Reply:
`MTHUB_ERROR` takes two arguments, so if the message is `"foobar"` then `OStrings.getString("MTHUB_ERROR", e.getMessage())` will result in `"Error foobar: "`. So you either need a dummy first argument, or you need a separate string template.

Reply:
it's true, I didn't realize it.
I add the template MTHUB_ERROR_MESSAGE=Error: {0}
and the call is OStrings.getString("MTHUB_ERROR_MESSAGE", e.getMessage());